pan3d/dataset_viewer.py

```diff
@@ -0,0 +1,485 @@
+        if "y" in kwargs and kwargs["y"] != self.state.render_y_scale:
+            self.state.render_y_scale = int(kwargs["y"])
+        if "z" in kwargs and kwargs["z"] != self.state.render_z_scale:
+            self.state.render_z_scale = int(kwargs["z"])
+        self.plotter.set_scale(
+            xscale=self.state.render_x_scale or 1,
+            yscale=self.state.render_y_scale or 1,
+            zscale=self.state.render_z_scale or 1,
+        )
+
+    def set_render_options(
+        self,
+        colormap: str = "viridis",
+        transparency: bool = False,
+        transparency_function: str = None,
+        scalar_warp: bool = False,
+    ) -> None:
+        """Set available options for rendering data.
+
+        Parameters:
+            colormap: A colormap name from Matplotlib (https://matplotlib.org/stable/users/explain/colors/colormaps.html)
+            transparency: If true, enable transparency and use transparency_function.
+            transparency_function: One of PyVista's opacity transfer functions (https://docs.pyvista.org/version/stable/examples/02-plot/opacity.html#transfer-functions)
+            scalar_warp: If true, warp the mesh proportional to its scalars.
+        """
+        if self.state.render_colormap != colormap:
+            self.state.render_colormap = colormap
+        if self.state.render_transparency != transparency:
+            self.state.render_transparency = transparency
+        if self.state.render_transparency_function != transparency_function:
+            self.state.render_transparency_function = transparency_function
+        if self.state.render_scalar_warp != scalar_warp:
+            self.state.render_scalar_warp = scalar_warp
+
+        if self.builder.mesh is not None and self.builder.data_array is not None:
+            self.plot_mesh()
+
+    def plot_mesh(self) -> None:
+        """Render current cached mesh in viewer's plotter."""
+        self.plotter.clear()
+        args = dict(
+            cmap=self.state.render_colormap,
+            clim=self.builder.data_range,
+            scalar_bar_args=dict(interactive=True),
+        )
+        if self.state.render_transparency:
+            args["opacity"] = self.state.render_transparency_function
+
+        try:
+            mesh = self.builder.mesh
+        except Exception as exception:
+            with self.state:
+                self.state.ui_error_message = str(exception)
+                self.state.ui_loading = False
+                return
+
+        if self.state.render_scalar_warp:
+            mesh = mesh.warp_by_scalar()
+        self.actor = self.plotter.add_mesh(
+            mesh,
+            **args,
+        )
+        if len(self.builder.data_array.shape) > 2:
+            self.plotter.view_isometric()
+        else:
+            self.plotter.view_xy()
+
+        if self.plot_view:
+            self.plot_view.push_camera()
+            self.plot_view.view_update()
+
+        with self.state:
```

Comment:
Do you need the `with state:`?
If so why it is not global to the method?

Reply:
Yes, the `with state` is still necessary since this is being executed asynchronously. Without it, the loading spinner doesn't go away when the render is complete. I can move this to encompass the whole method, though.

Reply:
Changed in [b87a945](https://github.com/Kitware/pan3d/pull/59/commits/b87a9450a989c0c672d41afe7d852b030323d219)